packages/client-sdk-nodejs/src/internal/internal-auth-client.ts

```diff
@@ -169,39 +184,152 @@ function tokenPermissionToGrpcPermission(
+    grpcPermission.topic.all =
+      new grpcAuth._GenerateApiTokenRequest.PermissionsType.All();
+  } else if (permission.topic instanceof TopicName) {
+    grpcPermission.topic.topic_name = permission.topic.name;
+  } else {
+    throw new Error(
+      `Unrecognized topic specification in topic permission: ${JSON.stringify(
+        permission
+      )}`
+    );
+  }
+  return grpcPermission;
 }
 
 function cachePermissionToGrpcPermission(
   permission: CachePermission
 ): grpcAuth._GenerateApiTokenRequest.PermissionsType.CachePermissions {
+  const grpcPermission =
+    new grpcAuth._GenerateApiTokenRequest.PermissionsType.CachePermissions();
   switch (permission.cacheRole) {
     case CacheRole.None:
       throw new Error('CacheRole.None not yet supported');
     case CacheRole.ReadWrite: {
-      const grpcPermission =
-        new grpcAuth._GenerateApiTokenRequest.PermissionsType.CachePermissions();
       grpcPermission.role =
         grpcAuth._GenerateApiTokenRequest.CacheRole.CacheReadWrite;
-      return grpcPermission;
+      break;
+    }
+    case CacheRole.ReadOnly: {
+      grpcPermission.role =
+        grpcAuth._GenerateApiTokenRequest.CacheRole.CacheReadOnly;
+      break;
+    }
+    default: {
+      throw new Error(`Unrecognized cache role: ${JSON.stringify(permission)}`);
     }
   }
 
-  throw new Error(
-    `Unrecognized cache permission: ${JSON.stringify(permission)}`
-  );
+  grpcPermission.cache =
+    new grpcAuth._GenerateApiTokenRequest.PermissionsType.CacheResource();
+  if (permission.cache instanceof All) {
+    grpcPermission.cache.all =
+      new grpcAuth._GenerateApiTokenRequest.PermissionsType.All();
+  } else if (permission.cache instanceof CacheName) {
+    grpcPermission.cache.cache_name = permission.cache.name;
+  } else {
+    throw new Error(
+      `Unrecognized cache specification in cache permission: ${JSON.stringify(
+        permission
+      )}`
+    );
+  }
+  return grpcPermission;
+}
+
+interface ResourceToPermissionMap {
+  [resource: string]: Permission;
+}
+
+function validateExplicitPermissions(permissions: Array<Permission>) {
+  if (!permissions || permissions.length < 1) {
+    throw new InvalidArgumentError('At least 1 permission must be specified.');
+  }
+  if (permissions.length > MAX_PERMISSIONS_PER_TOKEN) {
+    throw new InvalidArgumentError(
+      `A token cannot have more than ${MAX_PERMISSIONS_PER_TOKEN} permissions, ${permissions.length} permissions were provided.`
+    );
+  }
+
+  const cachePermissionMap: ResourceToPermissionMap = {};
+
+  function groupByCache(permission: CachePermission) {
+    const key = JSON.stringify(permission.cache);
+    const entry = cachePermissionMap[key];
+    if (entry instanceof CachePermission) {
+      // Cannot add multiple permissions for the same resource.
+      throw new InvalidArgumentError(
+        `Cache ${key} cannot have multiple permissions [${permission.cacheRole}, ${entry.cacheRole}]`
```

Comment:
do we want to do this validation in the client? i almost think this would be better to just do on the server so that the errors are consistent across the different SDKs and we don't have to repeat the code so much?  just a thought.

Reply:
I was going to add the validation in both client and server, but I can remove it from the client if we have been following that pattern for other apis.

Reply:
i wouldn't say it's a pattern, more of a case-by-case thing. but the more complex the validation is, the more concerned i am about the implementation drifting in one SDK vs the next vs the server. So I honestly might err towards just server-side on this one. but it's not a blocker for me

Reply:
Makes sense. Validation of this complexity is better off in the server.